Make sure enable! and disable! do not duplicate rules

mortenpi opened the pull request:
I believe this should fix #45.

However, in order to use `in` to determine if a particular rule is present in `p.rules`, I also have to add some overloads for `==` for the `*Rule`s that have some internal fields. I assume it is fine to consider two e.g. `FootnoteRules` equal, even in their caches differ? For now, I only added the overloads for `FootnoteRule` and `TableRule`, but if these overloads are fine, I can also add them to other rules that have fields.

mortenpi commented on src/parsers/rules.jl:
```diff
@@ -27,6 +27,7 @@ enable!(p::AbstractParser, fn, rules::Union{Tuple,Vector}) = (foreach(r -> enabl
 enable!(p::AbstractParser, fn, rule) = enable!(p, fn, fn(rule))
 
 function enable!(p::AbstractParser, rule)
+    rule ∈ p.rules && return p
```
I am not sure if this is actually the best behavior here. I wonder if it should instead either throw an error or replace an existing rule, should one already exist (thinking of the use case where you might want to override a rule).

MichaelHatherly replied:
Probably throw an error, we could add an `override!` that is meant for replacing existing rules and doesn't throw, that way it's a bit more explicit that that is the users' intension.

MichaelHatherly commented:
Unsure about overloading `==` here, perhaps it would make sense to just have a separate internal function for checking whether a rule already exists.

mortenpi commented:
> Unsure about overloading == here, perhaps it would make sense to just have a separate internal function for checking whether a rule already exists.

That also crossed my mind, will change to that.

mortenpi commented:
Updated. Added a function `is_same_rule` that could be overloaded if need be, but I assume the default implementation should work for all the current rules. Do you want me to add `override!` right away too?

MichaelHatherly commented on src/parsers/rules.jl:
```diff
@@ -11,6 +11,18 @@ struct Rule
+# disabling them in the parser) if the types match --- the values on any fields do not
+# matter. In case this is not correct for a rule, the two-argument is_same_rule should be
+# appropriately overloaded. Some possible cases where this might be necessary:
+#   (1) A rule with type parameters, where even when the type parameter values are
```
As a general note on this one, having rules that encode type parameters should be avoided where possible I think. I don't think any of the current set do, and from what I recall performance was worst generally when using `T` rather than just having `::Function` fields for example.

Fine to leave this note in place though.

mortenpi replied:
Hmm, interesting. I wonder if we're talking past each other here a little bit. What I mean in that note is something like `LinkRule{T}`, and then you'd `enable!` a fully qualified instance of it (i.e. `enable!(p, LinkRule{Foo}()`). I don't see how that would really affect performance, since I don't think `p.rules` is even really accessed during parsing?

I presume you might mean having the `Rule` type take type parameters?

MichaelHatherly replied:
> I presume you might mean having the Rule type take type parameters?

Either really, I'm not seeing a use case for where type parameters would really be needed for how the rules are meant to be defined/used. There may very well be one that I'm not seeing, but having the note there is fine by me. We can revisit should be come to a situation where type parameters are actually required somewhere to make things work nicely.

MichaelHatherly reviewed (approved):
Thanks, looks fine to me.
